corehq/ex-submodules/casexml/apps/phone/restore.py

```diff
@@ -193,7 +193,7 @@ def __init__(self, task, username):
         self.progress = {
             'done': task_info.get('done', 0),
             'total': task_info.get('total', 0),
-            'retry_after': task_info.get('retry-after', ASYNC_RETRY_AFTER),
+            'retry_after': RedisExponentialBackoff.get_next_time("%s.%s" % (task.id, username)),
```

Comment:
my only worry with this is that... this is going to increase exponentially.
Can we cap this to have a maximum retry-after timeout to be 30s or something? If this is too high, the phone will never poll the server for the sync response or a progress update, and the user will be stuck waiting for something to happen without any indication of what is going on.
Also, I think we should only incrementally back off after a certain amount of time has taken place (maybe a minute or so).

Reply:
That still doesn't sound like a bug to me. If anything, the UX should probably include some fakery rather than hitting our system every 5 seconds

Would a slower rate of expansion make sense? Like a factor of 1.5 rather than 2?

Reply:
I may have misunderstood what this is doing. What I understand is this:

Let's say your restore takes 8.5 minutes to generate, i.e. 510 seconds.
With your algorithm, the retry after times are as follows:

number | retry-after | cumulative
-- | -- | --
0 | 5 | 5
1 | 10 | 15
2 | 20 | 30
3 | 40 | 60
4 | 80 | 120
5 | 160 | 240
6 | 320 | 480
7 | 640 | 960
8 | 1280 | 1920
9 | 2560 | 3840
10 | 5120 | 7680

This means the phone will try and fetch the payload at 5s, 15s, 30s, etc. after it first requested it. This is fine for payloads that don't take very long. However, in the case where the payload takes 510s to generate, the user will actually have to wait 960s in total in order to receive the payload (as the phone will not ask the server for the payload between 480s and 960s). This is almost double the amount of time, and is making the problem of long restore times even worse. 

Did I understand that correctly?

The real way we should be doing this is estimating how long the restore might take and sending that down as the retry-after header. That was always the intention initially, but of course we never got around to implementing that part...

Reply:
Agreed that this needs a bit more brain storming.. 
Just throwing a thought here, feel free to neglect it if it does not make sense,

exponentially rise to a certain point say retry after `80` and then just keep it to a number higher than default but not high, say `10` .. so first increase from 5 till 80 and then just keep it 10.
This would help address both points,
1. the one being addressed by this PR
2. the point that Farid raised about restores that can take a while, so after say 80, it should keep checking on a regular interval considering that its a longer restore but since it has been in progress for a while, it should be ready any moment now..

can also consider cycling from 5 to 80 and then again 5 to 80 but might be a bit of overkill.

Reply:
Nailed it Farid. For my last comment I was starting to write something to generate that table, before thinking better of it.

I'm curious about the actual distribution of restore times. Do they usually take 10 minutes or more? What's the max amount of time a restore would generally take? And in those cases does the user sit and watch and wait the entire time until it happens?

Does aa6026e handle this comment reasonably?

Also, to be clear, as we discussed this pr doesn't handle the problem of long restores, but rather the thundering herd problem, i.e. when someone like rec decides to do 600 restores all at once we don't need to field 7200 extra requests per minute until they are all finished.